### ISSUE TITLE
Test: Use dev container in CI build

### ISSUE THREAD

**Issue by stuartleeks** [microsoft/vscode-remote-release#6769]
Refs: n/a

- [x] anyOS @amunger 
- [x] anyOS @alexr00

Complexity: 4

[Create Issue](https://github.com/microsoft/vscode-remote-release/issues/new?body=Testing+%236769%0A%0A&assignees=stuartleeks)

---

## Set up a repo to test
- Fork https://github.com/microsoft/vscode-remote-try-go
- In the fork (on GitHub), click on `.devcontainer` and then `devcontainer.json`
- Click the pencil icon to edit the file. At the end of the file, add the terraform feature:
```json
	"features": {
		"terraform": "1.1"
	}
```
- Add a commit message and commit directly to the main branch

## Set up dev container CI for the repo
- In the fork on GitHub, click "Actions" and "set up a workflow yourself"
- In the GitHub workflow YAML, add the following to the end of the sample workflow (and change the imageName value to match your repo name):
```yaml
      - name: Set up Docker BuildKit
        uses: docker/setup-buildx-action@v1

      - name: Login to GitHub Container Registry
        uses: docker/login-action@v1 
        with:
          registry: ghcr.io
          username: ${{ github.repository_owner }}
          password: ${{ secrets.GITHUB_TOKEN }}

      - name: Run build in dev container
        uses: stuartleeks/devcontainer-build-run@main
        with:
          # NOTE - Change the image name below to match your fork
          imageName: ghcr.io/stuartleeks/vscode-remote-try-go-devcontainer
          # build and show the contents of the folder (to show the compiled app)
          runCmd: |
            go build
            ls -al
            terraform -version
```
- Go to Actions and click on the latest action run
- When the workflow completes, expand the "Run build in dev container" output
- Under the "Run command in dev container", check that the directory listing shows the `vscode-remote-try-go` compilation output

## Test CI image caching
- Now re-run the workflow using the "Re-run all jobs" button above the build output
- When the workflow completes, expand the "Run build in dev container" output
- Expand the "build container" output and validate that the second execution used the cached container image from the first execution by checking for `CACHED` in the output, for example:
```
  #5 [2/2] RUN if [ "lts/*" != "none" ]; then su vscode -c ". /usr/local/share/nvm/nvm.sh && nvm install lts/* 2>&1"; fi
  #5 CACHED
```
- Also verify that the `terraform` CLI was installed by checking for the output of the `terraform -version` command:
```
Terraform v1.1.9
```
 
## Test building locally with VS Code

With the Remote-Containers extension and Docker installed:

- Clone the forked repo above to your local machine
- Open the cloned code in VS Code (but _do not_ re-open in Remote-Containers when prompted)
- Open `.devcontainer/devcontainer.json`  and add the `cacheFrom` under the `build` property (changing the image name to match your username as in the workflow YAML previously) as shown below:
```json
	"build": {
		"dockerfile": "Dockerfile",
		"args": {
			// Update the VARIANT arg to pick a version of Go: 1, 1.16, 1.17
			// Append -bullseye or -buster to pin to an OS version.
			// Use -bullseye variants on local arm64/Apple Silicon.
			"VARIANT": "1.17-bullseye",
			// Options
			"NODE_VERSION": "lts/*"
		},
		"cacheFrom": "ghcr.io/stuartleeks/vscode-remote-try-go-devcontainer"
	},
```
- From the command palette, select "Remote-Containers: Rebuild and Reopen in container"
- When the "Starting Dev Container (show log)" toast notification appears, click on it to bring up the Remote-Containers log
- In the log, check that the image built in CI was used as an image cache by searching for `CACHED`, for example:
```
CACHED [2/2] RUN if [ "lts/*" != "none" ]; then su vscode -c ". /usr/local/share/nvm/nvm.sh && nvm install lts/* 2>&1"; fi
```
- Open the Integrated Terminal and run `terraform -version` and verify that version 1.1 is reported

**Comment by amunger** [comment]
after reloading the window, the process automatically restarted and seems to have finished successfully this time. It looks like the cached image was used locally, not in the GH action, but I think you addressed that in your comment already

**Comment by chrmarti** [comment]
@alexr00 @amunger I have published a new version of @devcontainers/cli (0.4.0) with the fix for the caching issue. Thanks.
